### DISCUSSION TITLE
Is it possible to extend the framework capabilities from an external Module?

### DISCUSSION THREAD

**Discussion by elgransan** [oqtane/oqtane.framework#2157] [General]
I need to get all the users to allow to select one in my module but it isn't implemented in the core framework, Which is the better approach to add this functionality without touching the framework (the UserService in this case) ?

Thanks

**Comment by sbwalker** [comment]
@elgransan In order to answer this question, the first thing which needs to be determined is the security context of users who will be accessing the list of users in your custom module. Will they be anonymous users or authenticated users? If authenticated, what roles will they have?

**Comment by elgransan** [comment]
There are all authenticated users with custom roles. but one user must assign an item to an account manager (a user within another role), so I need to popup all the users from that role, I know it's not part of the core, and I could create a new repository inside my module (actually I did), I want to know if is it the correct approach?

**Comment by sbwalker** [comment]
@elgransan in regards to "one user must assign an item to an account manager"  - is that user in the account manager role as well?

**Comment by elgransan** [comment]
no, is in another role

**Comment by sbwalker** [comment]
The reason I am asking these questions is because in a client/server application, the API is publicly accessible so every endpoint needs to be secured appropriately. Currently the methods in the UserRoles API are only accessible by Administrators - to prevent unauthorized access to user information. This is a fairly high bar (which is the approach taken throughout the framework to ensure a least privilege model). So the question is, are there scenarios where a non-Administrator should be able to access user information? Certainly an unauthenticated user should not be able to retrieve a list of users for a site. But what about authenticated users? In general most APIs will not allow you to retrieve a full population of users even if you are authenticated. So what about a subset of users? If a user is a member of a role then it probably makes sense that they should be allowed to retrieve a list of the users in that role (the exception being the Registered Users role as ALL users are part of that role). But that does not solve your specific use case. So just to be clear... even if you implement your own endpoint in your custom module, I am trying to determine how it could be done in a way which does not compromise the security/privacy of the installation.

**Comment by elgransan** [comment]
ok I understand the security risks, you left me thinking, maybe I could solve this in another way. Thank you for the explanation and the time

**Reply by sbwalker** [reply]
@elgransan I appreciate you raising the question, as in taking a "least privilege" approach there may be some areas which are too "locked down" to make them practically usable for module developers. However, unlocking additional use cases should only be done with careful consideration of their impact. I do believe that it would be useful to introduce another UserRole GET method to retrieve the list of users in a specific role - as long as the caller is a member of the role (Registered Users role excluded). But like I said, this does not solve your scenario.

**Comment by elgransan** [comment]
Yes, that's why at first I wasn't asking to add this as part of the framework because is a specific request for a particular use case as you said. I think when you develop a entire app with several specific modules you could need something that isn't provided by the framework, and is ok because are special cases but in those cases is nice to have a standard way to extend the capabilities. And if those cases are creating security risks it is responsibility of the developer to address them.
